rfcs/convergence/input-native-element-props.md

```diff
@@ -0,0 +1,192 @@
+-->
+
+---
+
+@sopranopillow @ecraig12345
+
+## Summary
+
+This RFC proposes options on how to handle the native element props for input components on `vNext`. Currently the native props for all vNext components are applied to the root element, but in the case of an input component there are native input props that should be handled as well.
+
+<!-- Explain the proposed change -->
+
+## Problem statement
+
+For most components, it makes sense to apply native props (and the ref) to the root element.
+
+The input components are a little unique because the actual `<input>` *cannot* be the root element of the component, since we need various wrappers for styling. For example, rendering a `<Checkbox/>` actually does this:
+```tsx
+<div>
+  <div ... /> {/* checkmark */}
+  <input type="checkbox" />
+  <label {...slotProps.label} />
+</div>
+```
+Or an `<Input/>` is roughly like this:
+```tsx
+<div>
+  {/* optional bookend before slot */}
+  <div> {/* wrapper visually styled as an input (maybe, TBD) */}
+    {/* optional start slot */}
+    <input type="text" .../>
+    {/* optional end slot */}
+  </div>
+  {/* optional bookend after slot */}
+</div>
+```
+
+Another unique thing about inputs is that to the degree possible, we'd like them to work nicely with 3rd-party form libraries, which may have APIs that expect to be able to pass native props to the component and have them applied to the actual `<input>`. (needs more research)
+
+### [Formik](https://github.com/formium/formik)
+
+### [react-hook-form](https://github.com/react-hook-form/react-hook-form)
+
+Note, for this one they'd need to make a wrapper for our component because the library expects to be able to *set* `.value` on the input, which we probably don't intend to support. (Someone brought this up for v8 [here](https://github.com/microsoft/fluentui/issues/18126).)
+
+### [React Final Form](https://github.com/final-form/react-final-form)
+
+
+
+<!--
+Why are we making this change? What problem are we solving? What do we expect to gain from this?
+
+This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.
+
+In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
+-->
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+We don't have one suggested/preferred solution yet, because all of them have some significant problems.
+
+The examples below assume you're rendering a Checkbox and this is roughly your desired HTML output (it would also include a generated ID to associate the label and input):
+```html
+<div>
+  <input type="checkbox" name="foo" checked />
+  <label>sample</label>
+</div>
+```
+
+### Option 1: All native props always applied to root (current behavior)
```

Comment:
Could you do a survey of other react design systems and what solutions they use ? AFAIK most design systems will follow something along the lines of `Option 1`.

Reply:
I'd have to dig into it a bit more for Input (and Esteban would have a better idea for Checkbox) but I don't believe that's true. IIRC from the research I did, they appear to most commonly use either option 2 (combination) or possibly tending to apply the props just to the input not the root. But to confirm that I'd need to make a bunch of codesandboxes and look at their props and DOM.

Reply:
So far IMO I'd go for Option 1. seems more consistent with all other converged components.
But yeah, the arguments of counter intuitivity is valid.

Perhaps you could faltten some obvious properties in the root element and manually transfer it to the slot.

Reply:
So far IMO I'd go for Option 1. seems more consistent with all other converged components.
But yeah, the arguments of counter intuitivity is valid.

Perhaps you could faltten some obvious properties in the root element and manually transfer it to the slot.

Reply:
> you could faltten some obvious properties in the root element and manually transfer it to the slot.

unless this is consistent and explicitly stated for all vNext controls, I'd avoid any "special case" magic

Reply:
> Perhaps you could faltten some obvious properties in the root element and manually transfer it to the slot.

@bsunderhus Define "obvious"? I haven't been able to come up with a definition that relies on anything besides intuition, which is subjective and varies between people.

Reply:
Yeah, you're right. This is too much subjective.

Reply:
I checked a few other systems that I could readily get a working sandbox for their text Input component, and the consensus seemed to be something like Option 2, only slightly modified in that _all_ native props go to the input element, with a small number of exceptions that go to the wrapper (e.g. `className`).

I tested by adding a few props to the checkbox component and seeing what DOM element they ended up on (using `data-test` as a proxy for "any arbitrary prop"). The results were similar for most, but I've marked the differences with [*]

|  | name | id | className | aria-label | data-test |
|---|---|---|---|---|---|
| [Ant](https://ant.design/components/input/) | input | input | wrapper | input | input |
| [Carbon](https://www.carbondesignsystem.com/components/text-input/code/) | input | input | input [*] | input | input |
| [Fabric v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield) | input | input | wrapper | - | input |
| [Northstar v0](https://fluentsite.z22.web.core.windows.net/0.57.0/components/input/definition) | input | input | wrapper | input | wrapper  [*] |

* Carbon put `className` on the input element for text Input component in that table, but for the Checkbox component, it put `className` on the label element (not the wrapper, nor the input elements).